File: prova.cpp

int main()
{
    int x = 10;
    int y = 5;
    int z = 10;

    void pippo(int y, int *z)
    {
        z = x + y + z;
    }

    {
        int x = 20;
        int y = 30;
        int z = 50;
        pippo(x++, x);
        pippo(x++, x);
        printf("%d", x);
    }
    printf("%d", x);
}

File: prova.pas
program Example;

var
  x: Integer;

procedure ass_x(n: Integer);
begin
  x := n;
end;

procedure stampa_x;
begin
  WriteLn(x);
end;

procedure pippo(S: Procedure; P: Procedure; n: Integer);
var
  x: Integer;
begin
  x := 10;
  if n = 1 then
    ass_x(n)
  else
    S(n);
  P;
  stampa_x;
end;

begin
  x := 3;
  pippo(@ass_x, @stampa_x, 1);
  pippo(@ass_x, @stampa_x, 2);
end.